internal/uvm/virtual_device.go

```diff
@@ -15,54 +15,106 @@ import (
+	if err := vpci.vm.RemoveDeviceFromUVM(ctx, vpci); err != nil {
 		log.G(ctx).WithError(err).Warn("failed to remove VPCI device")
 		return err
 	}
 	return nil
 }
 
-// AssignDevice assigns a new vpci device to the uvm
-func (uvm *UtilityVM) AssignDevice(ctx context.Context, device hcsschema.VirtualPciDevice) (*VPCIDevice, error) {
+// AssignDeviceToUVM assigns a vpci device to the uvm
+// if the device already exists, the stored VPCIDevice's ref count is increased
+// and the VPCIDevice is returned.
+// Otherwise, a new request is made to assign the target device indicated by the deviceID
+// onto the UVM. A new VPCIDevice entry is made on the UVM and the VPCIDevice is returned
+// to the caller
+func (uvm *UtilityVM) AssignDeviceToUVM(ctx context.Context, deviceID string) (*VPCIDevice, error) {
 	guid, err := guid.NewV4()
 	if err != nil {
 		return nil, err
 	}
-	id := guid.String()
+	vmBusGUID := guid.String()
 
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	if err := uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
+
+	existingVPCIDevice := uvm.vpciDevices[deviceID]
+	if existingVPCIDevice != nil {
+		existingVPCIDevice.refCount++
+		return existingVPCIDevice, nil
+	}
+
+	targetDevice := hcsschema.VirtualPciDevice{
+		Functions: []hcsschema.VirtualPciFunction{
+			{
+				DeviceInstancePath: deviceID,
+			},
+		},
+	}
+
+	request := &hcsschema.ModifySettingRequest{
+		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, vmBusGUID),
 		RequestType:  requesttype.Add,
-		Settings:     device,
-		GuestRequest: guestrequest.GuestRequest{
+		Settings:     targetDevice}
+
+	// we do not currently support assigned devices on windows,
+	// so no guest side work necessary for windows
+	if uvm.operatingSystem != "windows" {
+		// for LCOW, we need to make sure that specific paths relating to the
+		// device exist so they are ready to be used by later
+		// work in openGCS
+		request.GuestRequest = guestrequest.GuestRequest{
 			ResourceType: guestrequest.ResourceTypeVPCIDevice,
 			RequestType:  requesttype.Add,
 			Settings: guestrequest.LCOWMappedVPCIDevice{
-				VMBusGUID: id,
+				VMBusGUID: vmBusGUID,
 			},
-		},
-	}); err != nil {
+		}
+	}
+
+	if err := uvm.modify(ctx, request); err != nil {
 		return nil, err
 	}
-	return &VPCIDevice{
-		vm: uvm,
-		ID: id,
-	}, nil
+	result := &VPCIDevice{
+		vm:               uvm,
+		VMBusGUID:        vmBusGUID,
+		deviceInstanceID: deviceID,
+		refCount:         1,
+	}
+	uvm.vpciDevices[deviceID] = result
+	return result, nil
 }
 
-// RemoveDevice removes a vpci device from the uvm
-func (uvm *UtilityVM) RemoveDevice(ctx context.Context, id string) error {
+// RemoveDeviceFromUVM removes a vpci device from a uvm when there are
+// no more references to a given VPCIDevice. Otherwise, decrements
+// the reference count of the VPCIDevice and returns nil.
+func (uvm *UtilityVM) RemoveDeviceFromUVM(ctx context.Context, vpci *VPCIDevice) error {
```

Comment:
I think it's okay to keep it so this function takes the instance ID rather than the entire `VPCIDevice`. It looks a little confusing to me this way wondering whether we need to do extra work to ensure the passed in device matches what we find in the map or not, and we could avoid that by just taking an ID and looking it up.

If possible, I would like to make this function non-exported anyways, and treat it as simply an implementation detail on `VPCIDevice.Release`, rather than something another package could call into directly.